blog/_posts/2019-09-06-profilers.md

```diff
@@ -0,0 +1,405 @@
+
+![PProf](/images/blog/2019-09-06-profilers/profile-pprof.png)
+
+This is nice! We get to visualize the static functions call tree simultaneously with the weight each function contributes to the time. However, this format can be a bit limited on the total quantity of information it can fit, so it might not always help. The `pprof` tool automatically filters out smaller frames to aid the user in seeing the important ones, and provides several options for adding and removing functions from the view.
+
+In addition, the `pprof` tool has a sandwich text report that summarizes the caller and callee list around each profile sample:
+
+```
+      Type: events
+Showing nodes accounting for 632, 100% of 632 total
+----------------------------------------------------------+-------------
+      flat  flat%   sum%        cum   cum%   calls calls% + context
+----------------------------------------------------------+-------------
+                                               273 52.20% |   fib REPL:-1
+                                               273 52.20% |   fib_r REPL:1
+       417 65.98% 65.98%        523 82.75%                | fib REPL:2
+                                               273 52.20% |   fib_r REPL:1
+----------------------------------------------------------+-------------
+                                               273 72.03% |   fib REPL:2
+       178 28.16% 94.15%        379 59.97%                | fib_r REPL:1
+                                               273 72.03% |   fib REPL:2
+                                                33  8.71% |   + ./int.jl:2
+                                                 1  0.26% |   - ./int.jl:2
+----------------------------------------------------------+-------------
+                                                33   100% |   + ./int.jl:2 (inline)
+        33  5.22% 99.37%         33  5.22%                | + ./int.jl:53
+----------------------------------------------------------+-------------
+                                                 1   100% |   - ./int.jl:2 (inline)
+         1  0.16% 99.53%          1  0.16%                | - ./int.jl:52
+----------------------------------------------------------+-------------
+```
+
+Let's combine all those ideas to make a new option for our Julia plain-text tree view: flattened recursion.
+
+  * recur – Controls the recursion handling in `:tree` format. `off` (default) prints the tree as normal. `flat` instead compresses any recursion (by ip), showing the approximate effect of converting any self-recursion into an iterator. `flatc` does the same but also includes collapsing of C frames (may do odd things around `jl_apply`).
+
+This is no longer a literal interpretation of the stack, but shows a particular linearization where each recursive call was re-written as a loop.
```

Comment:
There's not quite enough information given to understand the rules here. I think what you mean is that if an IP appears twice in the trace, you truncate the trace at the second one. But why the second one? Why not the first one?

Reply:
Neither. It's not truncated, it's combined.

Fun fact: the new `format=:flat` algorithm forms a subset of this new `tree-flat` option, but where all calls (instead of just recursive ones) are reinterpreted as if called from a single interpreter loop. To put that in the context of Julia, if you had a tree of calls that all went through `jl_apply`, the `tree-flatc` and `format=flat` output should start to coincide (plus-or-minus some formatting and indentation).